### PR TITLE
Topography meshing

### PR DESCRIPTION
**Pull request by dclenz**
New version of edge-cut, which improves efficiency and flexibility.

### CODE REVIEW

File: tools/edge_cut/src/main.cpp [hunk 1]
```diff
@@ -50,55 +70,121 @@ int main( int i_argc, char *i_argv[] ) {
...
+                                  CGAL::parameters::facet_size = l_bdryFacetCrit,
+                                  CGAL::parameters::facet_distance = l_bdryApproxCrit,
+                                  CGAL::parameters::facet_angle = l_angleBound         );
+
+  // Mesh generation
+  // NOTE the optimizers have more options than specified here
+  EDGE_LOG_INFO << "Re-meshing polyhedral surfaces according to provided criteria";
+  C3t3 topoComplex = CGAL::make_mesh_3<C3t3>( l_topoDomain,
+                                              l_topoCriteria,
+                                              CGAL::parameters::lloyd(    CGAL::parameters::time_limit = 60 ),
```
Comment (breuera) [inline tools/edge_cut/src/main.cpp]: What is ``CGAL::parameters::time_limit`` doing? Should be a runtime-parameters, for example, be part of a <cgal/> section in the config.

File: tools/edge_cut/src/surf/BdryTrimmer.h [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote products derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @section DESCRIPTION
+ * Delete extraneous bits of mesh which extend above topographical surface
+ **/
+
+#ifndef BDRY_TRIMMER_H_
```
Comment (breuera) [inline tools/edge_cut/src/surf/BdryTrimmer.h]: Add ``EDGE_CUT_`` or similar to all include guards, here ``EDGE_CUT_BDRY_TRIMMER_H_``.

File: tools/edge_cut/src/surf/BdryTrimmer.h [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+#define BDRY_TRIMMER_H_
+
+// #include "meshUtils.h"
```
Comment (breuera) [inline tools/edge_cut/src/surf/BdryTrimmer.h]: Remove, no dead code.

File: tools/edge_cut/src/surf/BdryTrimmer.inc [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+ **/
+
+template< class Polyhedron >
+edge_cut::surf::BdryTrimmer<Polyhedron>::BdryTrimmer( Polyhedron& i_bdry, Polyhedron& i_topo ) :
+  m_bdry(i_bdry),
+  m_topo(i_topo)
+{
+  m_bdry.normalize_border();
+
+  assert( !m_topo.is_empty() );
```
Comment (breuera) [inline tools/edge_cut/src/surf/BdryTrimmer.inc]: Replace all ``assert([...])`` by respective wraps of the logger.
For example: ``EDGE_CHECK``, ``EDGE_CHECK_NE``, ``EDGE_CHECK_EQ``.

File: tools/edge_cut/src/surf/BdryTrimmer.inc [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        l_vIt++ )
+  {
+    if ( l_vIt->point() == i_v->point() ) break;
+  }
+
+  if ( l_vIt == m_bdry.vertices_end() )
+  {
+    EDGE_LOG_INFO << "Could not find coincident vertex between topography and border mesh - mesh will not be trimmed";
+    EDGE_LOG_INFO << "Missing point: " << i_v->point();
+    assert( l_vIt != m_bdry.vertices_end() );
```
Comment (breuera) [inline tools/edge_cut/src/surf/BdryTrimmer.inc]: Use ``EDGE_LOG_FATAL`` for aborting the code.

File: tools/edge_cut/src/surf/BdryTrimmer.inc [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+  //         (v3) .    / U      [below topo]
+  //             \    /
+  //            W \  /
+  // ______________\/   (v1)
+  //
+  // Consider vector u = v_2 - v_1   and   w = v_3 - v_1
+  // We will compute the perpendicular vector from u into v_3 : if the z
+  // component of this vector is positive, then we are "above" the topography.
+  // Otherwise, we are below, and have to flip the halfedge to the other side.
+  double l_dotProdUW = (v3.x()-v1.x()) * (v2.x()-v1.x()) + (v3.y()-v1.y()) * (v2.y()-v1.y()) + (v3.z()-v1.z()) * (v2.z()-v1.z());
```
Comment (breuera) [inline tools/edge_cut/src/surf/BdryTrimmer.inc]: Is precision a template in CGAL. In that case, use the respective spec over ``double``?

File: tools/edge_cut/src/surf/Topo.cpp [hunk 7]
```diff
@@ -127,3 +165,44 @@ unsigned short edge_cut::surf::Topo::interSeg( CGAL::Point_3< CGAL::Cartesian<do
...
+  // write the vertices
+  std::map<Vertex_handle, std::size_t> index_of_vertex;
+
+  std::size_t v_idx = 0;
+  for( Vertex_iterator it = m_delTria->finite_vertices_begin(); it != m_delTria->finite_vertices_end(); ++it, ++v_idx )
+  {
+      os << *it << std::endl;
+      index_of_vertex[it] = v_idx;
+  }
+  CGAL_assertion( v_idx == num_verts );
```
Comment (breuera) [inline tools/edge_cut/src/surf/Topo.cpp]: Replace CGAL_assertions with native ones, e.g.,``EDGE_CHECK_EQ``

File: tools/edge_cut/src/surf/meshUtils.cpp [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+    // Control should never reach here if inner and outer radii are equal
+    assert( m_outerRad != m_innerRad );
+    return ( 1 + ( l_distance - m_innerRad ) * ( m_scale - 1) * m_widthInv ) * m_innerVal;
+  }
+}
+
+
+edge_cut::surf::Polyline_type
+edge_cut::surf::topoIntersect(  Poly_slicer & i_slicer,
+                                K::Plane_3    i_plane   )
```
Comment (breuera) [inline tools/edge_cut/src/surf/meshUtils.cpp]: Is ``K::Plane_3`` a pointer or should it be passed by reference as well?

File: tools/edge_cut/src/surf/meshUtils.cpp [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+  Polylines     polylines;
+  Polyline_type ridges;
+
+  i_slicer( i_plane, std::back_inserter( polylines ) );
+
+  if( polylines.size() == 1 ) {
+    ridges = *(polylines.begin());
+  }
+  else {
+    EDGE_LOG_ERROR << "Error: Polyline slicer did not return precisely one intersection!";
```
Comment (breuera) [inline tools/edge_cut/src/surf/meshUtils.cpp]: ``EDGE_LOG_ERROR`` does not abort the code, is this the intended behavior.
``EDGE_LOG_FATAL`` does.

File: tools/edge_cut/src/surf/meshUtils.cpp [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+}
+
+
+std::list< edge_cut::surf::Polyline_type >
+edge_cut::surf::getIntersectionFeatures(  Polyhedron  const & i_topoSurface,
+                                          double      const * i_bBox        )
+{
+  std::list< Polyline_type > features;
+
+  EDGE_LOG_INFO << "Computing topography-boundary intersection...";
```
Comment (breuera) [inline tools/edge_cut/src/surf/meshUtils.cpp]: Don't think we should have an info-message here (silent functions).
Move to message outside.

File: tools/edge_cut/src/surf/meshUtils.cpp [hunk 11]
```diff
@@ -0,0 +1,293 @@
...
+                                                                l_polygons,
+                                                                io_bdry      );
+  io_bdry.normalize_border();
+
+  return io_bdry;
+}
+
+
+void
+edge_cut::surf::c3t3ToPolyhedron( C3t3        const & c3t3,
```
Comment (breuera) [inline tools/edge_cut/src/surf/meshUtils.cpp]: Add ``i_``, ``io_``, ``o_`` etc to args.

**Comment by dclenz** [comment]
Thanks for the notes, I've begun to incorporate the suggestions.

About Valgrind: the precision used by CGAL does not play well with Valgrind, and there seems to be no intention of resolving this.  The result is that Valgrind can't be used with CGAL code (as far as I can tell).  It was addressed here: https://github.com/CGAL/cgal/issues/3094#issuecomment-389785782

For memory usage statistics I've used the gperftools and tcmalloc libraries (which you noted above, and I can address separately).

**Comment by breuera** [comment]
Ok, great.

Valgrind:
I see, this is unfortunate.
Can we run Clang's sanitizers? https://github.com/google/sanitizers
Has to be done on local machines though (supercomputers typically limit the virtual memory).

**Pull request merged**